### ISSUE TITLE
New well designed result printer mechanism

### ISSUE THREAD

**Issue by JeroenDeDauw** [SemanticMediaWiki/SemanticMediaWiki#125]
Right now the query results are tightly bound to the store, making it ~~nearly~~ impossible to test logic in consumers (ie result printers).
- [Original tread on the dev list](http://sourceforge.net/mailarchive/forum.php?thread_name=CAMhmagA8YDPw6mv%3D2NV1Mpn8qvS8J_LTHGdo1G4hWzPQ_idZ6Q%40mail.gmail.com&forum_name=semediawiki-devel)
- [Some conceptual hacking](https://github.com/JeroenDeDauw/CheckedTable/tree/master/src/SMW/Query/Result)

This item needs to be tackled before we can make big changes affecting result printers, such as adoption of the new data values libraries.

**Comment by mwjames** [comment]
Nothing has been done to resolve this task within the last 2 years, moving this two [0] where interested parties can still pick this up.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/projects/4

**Comment by JeroenDeDauw** [comment]
I was about to create a new issue to talk about an experiment I am doing to create a sane representation of a query result. Where sane includes:

* No services like the SMW Store required for construction
* No services exposed
* Immutable
* Good naming
* Dedicated objects and not arrays in arrays

GitHub showed me this issue. And surprise, my experiment came up with the exact same structure as I apparently already found 6 years ago. New experiment: https://github.com/ProfessionalWiki/ModernTimeline/tree/master/src/ResultFacade

Modifying the existing result printer interface is a big breaking change, and modifying it this much would be a ton of work. Plus likely a bunch of printers depend on complexity that is better avoided for others. So I suggest we create a new much simpler interface that works for 95% ish of formats. Such an interface would not have these methods https://github.com/ProfessionalWiki/ModernTimeline/blob/530bf26052aa8d3a2e33ab487c4f12523f24bf15/src/ModernTimelinePrinter.php#L94-L118 and be invoked with a query result like in the two experiments.

**Comment by JeroenDeDauw** [comment]
Progress: https://github.com/ProfessionalWiki/ModernTimeline/tree/37e16eb77bb7ea6972fd7fd869c072b16b2ddaec/src/ResultFacade

This now includes a draft for the new interface fulfilling these design improvements:

* Dependency injection possible
* Simple query result object
* Query result can be extended later without breaking changes (both data and logic)
* No exposure to legacy ParamProcessor classes
* Commonly needed info (name and parameters) split from printer instance to avoid not needed object graph instantiation

TODO:

* Pass something for side effects (like adding JS to otuput) to `getResult`
* `PrintRequest` has not been dealth with yet (global access and constructor parameters)
* naming naming naming
* Moving this into SMW

**Comment by JeroenDeDauw** [comment]
Code now available as PR, which also tracks remaining TODOs: https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/4313
